File: 20_valid_parentheses/20_valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        dict_rule = {
            '}': '{',
            ']': '[',
            ')': '(' 
        }

        for c in s:
            if c in dict_rule:
                if stack and stack[-1] == dict_rule[c]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)
        return len(stack) == 0 

File: 20_valid_parentheses/20_valid_parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  let ref = {
    "{": "}",
    "(": ")",
    "[": "]",
  };
  let res = [];
  for (let i = 0; i < s.length; i++) {
    if (ref[s[i]]) res.push(s[i]);
    else {
      if (res.length == 0 || ref[res.pop()] !== s[i]) {
        return false;
      }
    }
  }
  return res.length === 0;
};
